karamel-core/src/main/java/se/kth/karamel/backend/machines/SshMachine.java

```diff
@@ -183,50 +191,72 @@ private void runTask(Task task) {
-
-      session = client.startSession();
-      Session.Command cmd = session.exec(shellCommand.getCmdStr());
-      cmd.join(60 * 24, TimeUnit.MINUTES);
-      updateHeartbeat();
-      if (cmd.getExitStatus() != 0) {
-        shellCommand.setStatus(ShellCommand.Status.FAILED);
-      } else {
-        shellCommand.setStatus(ShellCommand.Status.DONE);
-        if (task instanceof RunRecipeTask) {
-          RunRecipeTask rrt = (RunRecipeTask) task;
+
+    while (!finished && numRetries > 0) {
+      shellCommand.setStatus(ShellCommand.Status.ONGOING);
+      try {
+        logger.info(machineEntity.getId() + " => " + shellCommand.getCmdStr());
+
+        session = client.startSession();
+        Session.Command cmd = session.exec(shellCommand.getCmdStr());
+        cmd.join(60 * 24, TimeUnit.MINUTES);
+        updateHeartbeat();
+        if (cmd.getExitStatus() != 0) {
+          shellCommand.setStatus(ShellCommand.Status.FAILED);
+          // Retry just in case there was a network problem somewhere on the server side
+        } else {
+          shellCommand.setStatus(ShellCommand.Status.DONE);
+          finished = true;
+          if (task instanceof RunRecipeTask) {
+            RunRecipeTask rrt = (RunRecipeTask) task;
+            try {
+              processReturnValues(rrt.getCookbookName(), rrt.getRecipeName());
+            } catch (JsonParseException p) {
+              logger.error("Bug in Chef Cookbook - Return JSON was not a valid json document from: "
+                  + rrt.getCookbookName() + "::" + rrt.getRecipeCanonicalName());
+            } catch (IOException e) {
+              logger.info("No return values for: "
+                  + rrt.getCookbookName() + "::" + rrt.getRecipeCanonicalName());
+            }
+          }
+        }
+        SequenceInputStream sequenceInputStream = new SequenceInputStream(cmd.getInputStream(), cmd.getErrorStream());
+        LogService.serializeTaskLog(task, machineEntity.getPublicIp(), sequenceInputStream);
+      } catch (ConnectionException | TransportException ex) {
+        if (getMachineEntity().getGroup().getCluster().getPhase() != ClusterRuntime.ClusterPhases.PURGING) {
+          logger.error(String.format("Couldn't excecute command on client '%s' ", machineEntity.getId()), ex);
+        }
+      } finally {
+        // Retry if we have a network problem
+        numRetries--;
+        if (!finished) {
           try {
-            processReturnValues(rrt.getCookbookName(), rrt.getRecipeName());
-          } catch (JsonParseException p) {
-            logger.error("Bug in Chef Cookbook - Return JSON was not a valid json document from: "
-                + rrt.getCookbookName() + "::" + rrt.getRecipeCanonicalName());
-          } catch (IOException e) {
-            logger.error("No return values for: "
-                + rrt.getCookbookName() + "::" + rrt.getRecipeCanonicalName());
+            Thread.sleep(timeBetweenRetries);
+          } catch (InterruptedException ex) {
```

Comment:
Read the line below. If it is stopping, it will rethrow the exception. That is still highest priority. If the user asked to stop/pause it will. However, it's some other cause for InterruptedException, it will retry.